### DISCUSSION TITLE
Are success messages translatable ?

### DISCUSSION THREAD

**Discussion by ryssbowh** [solspace/craft-freeform#911] [Questions & Answers] [answered]
**Describe the bug or issue you're experiencing**

I have multiple forms used on several sites for different languages but I can't seem to be able to translate the success messages.
I have a plugin to translate strings with the `site` and `freeform` categories enabled, and those strings aren't picked up.
I haven't tried but probably it's the same for error messages.

**Expected behavior**

Those strings are rendered using the `t` function (probably in the `site` category) so that messages can be translated in different languages.

**Craft & Plugin Info (please complete the following information):**

- Craft Version: 4.5.10
- Freeform Version: 4.1.11
- Freeform Edition: Pro
- Fresh Install or Upgrade: Fresh

**Comment by kjmartens** [comment]
Hi @ryssbowh,

I would just like to confirm if you have checked out the Translation guide here:
https://docs.solspace.com/craft/freeform/v4/setup/translating/

The success and error banner (when using AJAX) is controlled with a JS override:
https://docs.solspace.com/craft/freeform/v4/templates/ajax-forms/#customize-success-error-messages

``` twig
// Change success and error messages for all forms on this page
document.addEventListener('freeform-ready', function(event) {
    // Customize the error and success messages
    event.options.successBannerMessage = 'My custom success message';
    event.options.errorBannerMessage = 'My custom error message';
});
```

🙂

**Reply by ryssbowh** [reply]
Hi @kjmartens 

Thanks for the link, I had missed it.

I'm still unsure how I would go about this though. If I'm correct the success message should be translated before rendering, so I tried this :
```
{{ form.renderTag({
  successMessage: form.successMessage|t
}) }}
```
which doesn't work.

Now yes I could replace the success message in JS but that becomes quickly cumbersome; need to load the translated message in the template somehow, add custom javascript on the page for each different form etc.

Would it possible to allow overriding the success message as per my test above ?
Or translate it directly in your code on the `site` category ? I don't know all the implications but that seems a fairly safe thing to do ?

Edit: or allow overriding the success message : `{% do form.setSuccessMessage(form.successMessage|t) %}` or an event before render so they can be replaced in PHP. Anything to avoid having to have custom JS for something as simple as translating a string.

**Reply by ryssbowh** [reply, marked as answer]
Actually this is possible through an event :
```
Event::on(FormsService::class, FormsService::EVENT_ATTACH_FORM_ATTRIBUTES, function ($e) {
    $e->attachAttribute('data-success-message', \Craft::t('site', $e->form->getSuccessMessage()));
});
```
Which still feels cumbersome to me, but if you're happy with it this conversation can be closed

**Reply by kjmartens** [reply]
Yes, that's likely the way to go about it for now. We plan to revisit and improve this in the future though. 🙂
